apps/hermes/client/js/src/HermesConnection.ts

```diff
@@ -0,0 +1,198 @@
+import { EncodingType, PriceFeedMetadata, PriceUpdate } from "./types";
+import EventSource from "eventsource";
+
+const DEFAULT_TIMEOUT: DurationInMs = 5000;
+const DEFAULT_HTTP_RETRIES = 3;
+
+export type UnixTimestamp = number;
+export type DurationInSeconds = number;
+export type HexString = string;
+export type DurationInMs = number;
+
+export type HermesConnectionConfig = {
+  /* Timeout of each request (for all of retries). Default: 5000ms */
+  timeout?: DurationInMs;
+  /**
+   * Number of times a HTTP request will be retried before the API returns a failure. Default: 3.
+   *
+   * The connection uses exponential back-off for the delay between retries. However,
+   * it will timeout regardless of the retries at the configured `timeout` time.
+   */
+  httpRetries?: number;
+};
+
+export class HermesConnection {
+  private baseURL: string;
+  private timeout: DurationInMs;
+  private httpRetries: number;
+
+  /**
+   * Constructs a new Connection.
+   *
+   * @param endpoint endpoint URL to the price service. Example: https://website/example/
+   * @param config Optional HermesConnectionConfig for custom configurations.
+   */
+  constructor(endpoint: string, config?: HermesConnectionConfig) {
+    this.baseURL = endpoint;
+    this.timeout = config?.timeout || DEFAULT_TIMEOUT;
+    this.httpRetries = config?.httpRetries || DEFAULT_HTTP_RETRIES;
+  }
+
+  private async httpRequest<ResponseData>(
+    url: string,
+    options?: RequestInit,
+    retries = this.httpRetries,
+    backoff = 300
+  ): Promise<ResponseData> {
+    const controller = new AbortController();
+    const { signal } = controller;
+    options = { ...options, signal }; // Merge any existing options with the signal
+
+    // Set a timeout to abort the request if it takes too long
+    const timeout = setTimeout(() => controller.abort(), this.timeout);
+
+    try {
+      const response = await fetch(url, options);
+      clearTimeout(timeout); // Clear the timeout if the request completes in time
+      if (!response.ok) {
+        throw new Error(`HTTP error! status: ${response.status}`);
+      }
+      return await response.json();
+    } catch (error) {
+      clearTimeout(timeout);
+      if (
+        retries > 0 &&
+        !(error instanceof Error && error.name === "AbortError")
+      ) {
+        // Wait for a backoff period before retrying
+        await new Promise((resolve) => setTimeout(resolve, backoff));
+        return this.httpRequest(url, options, retries - 1, backoff * 2); // Exponential backoff
```

Comment:
This isn't exponential, this is linear.  To make it exponential this should be `Math.pow(2, this.httpRetries - retries) * initialBackoff`.

I would also suggest making the initial backoff a random number within a range of a few hundred ms, as this will help prevent a "thundering herd" problem where some network outage causes all clients to retry connecting at exactly the same moment, and then if opening the connections fails, they all back off at the exact same rate and so the backoff doesn't actually help.